tests/functional/src/main/java/org/mule/tck/junit4/ExtensionsFunctionalTestCase.java

```diff
@@ -0,0 +1,283 @@
+ * performed by invoking {@link ExtensionManager#discoverExtensions(ClassLoader)}.
+ * Although this behavior suits most use cases, it can be time consuming because of
+ * all the classpath scanning and the overhead of initialising extensions that
+ * are most likely not used in this tests. As the number of extensions available grows,
+ * the problem gets worst. For those cases,  you can override the {@link #getDescribers()}
+ * and specify which describers are to be used to initialise the extensions manager. In that way,
+ * extensions discovery is skipped and you only initialise what you need.
+ * <p/>
+ * Once extensions are discovered and described,
+ * a {@link ResourcesGenerator} is used to automatically
+ * generate any backing resources needed (for example, XSD schemas, spring bundles,
+ * service registration files, etc).
+ * <p/>
+ * In this way, the user experience is greatly simplified when running the test
+ * either through an IDE or build tool such as maven or gradle.
+ * <p/>
+ * Since this class extends {@link FunctionalTestCase}, a new {@link MuleContext}
+ * is created per each test. That also means that a new {@link ExtensionManager}
+ * is created per test.
+ *
+ * @since 3.7.0
+ */
+public abstract class ExtensionsFunctionalTestCase extends FunctionalTestCase
+{
+
+    private final ServiceRegistry serviceRegistry = new SPIServiceRegistry();
+    private final ExtensionFactory extensionFactory = new DefaultExtensionFactory(serviceRegistry);
+    private ExtensionManager extensionManager;
+    private File generatedResourcesDirectory;
+
+
+    @Override
+    protected void doSetUpBeforeMuleContextCreation() throws Exception
+    {
+        super.doSetUpBeforeMuleContextCreation();
+        createExtensionsManager();
+    }
+
+    /**
+     * Implement this method to limit the amount of extensions
+     * initialised by providing the {@link Describer}s for
+     * the extensions that you actually want to use for this test.
+     * Returning a {@code null} or empty array will cause the
+     * {@link #getAnnotatedExtensionClasses()} method to be considered.
+     * Default implementation of this method returns {@code null}
+     */
+    protected Describer[] getDescribers()
+    {
+        return null;
+    }
+
+    /**
+     * Implement this method to limit the amount of extensions
+     * initialised by providing the annotated classes which define
+     * the extensions that you actually want to use for this test.
+     * Returning a {@code null} or empty array forces the
+     * {@link ExtensionManager} to perform a full classpath discovery.
+     * Default implementation of this method returns {@code null}.
+     * This method will only be considered if {@link #getDescribers()}
+     * returns {@code null}
+     */
+    protected Class<?>[] getAnnotatedExtensionClasses()
+    {
+        return null;
+    }
+
+    @Override
+    protected final void addBuilders(List<ConfigurationBuilder> builders)
```

Comment:
Ass this class is intended to be extended, add javadocs to all protected methods